Implement RetryingClient

martinnj opened the pull request:
Goal here is basically to help with https://github.com/pinterest/pymemcache/issues/307.
And takes inspiration from this [comment](https://github.com/pinterest/pymemcache/issues/307#issuecomment-862109935)

The idea is a thin "wrapper client" that does nothing but forward calls and retry as appropriate.
The features are as follows:
 - Takes any Client or Client-like class.
 - Configurable number of retries.
 - Configurable sleep between retries. (Don't think it's really a good idea in a cache, but it was easy, so here we are.)
 - Allows retrying only for specific Exception types.
 - Allows skipping retries for specific Exception types.

I haven't written tests yet since I wanted some opinions on the interface/constructor first.

Currently the class uses `__getattr__` to pass through any attributes/methods that isn't present on itself to the inner client.
It also passes through `dir()` calls to the inner client to allow for checking available methods, attributes, etc.
Is this an acceptable implementation or is it too opaque?

I also did a more naive implementation, it is commented out in the code.

Feedback and suggestions are welcome.
I'll write the unit-tests ones the feel of the client is more solid.

@cgordon are you the person to go to for this? 😄

cgordon commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,297 @@
+        for attempt in range(self.attempts):
+            try:
+                return self.client.__getattribute__(name, *args, **kwargs)
+            except Exception as exc:
```
Have a special case for `AttributeError` in case the user calls a function that isn't defined on the underlying `client`.

martinnj replied:
Just to be sure, you mean catch `AttributeError`?
Propegating it seems like the right call since you've essentially tried to call a specific method from the underlying `client`, that method doesn't exist. Which would also give the `AttributeError` if called directly.

Or did I misunderstand the comment? :)

cgordon replied:
We need to make sure to never retry on an `AttributeError`.

cgordon replied:
Of course, a user of the library should be able to retry on their own `AttributeError`s! So we need logic that ignores AttributeError thrown by our code only.

martinnj replied:
Gotcha!

martinnj replied:
Redid the retry logic so this isn't relevant anymore. :)

cgordon commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,297 @@
+    Client that allows retrying calls for the other clients.
+    """
+
+    def __init__(
```
Document the arguments to this method. For example, what does it mean for `retry_for` to be `None`? Does it mean that all exceptions cause retries, or no exceptions cause retries?

cgordon commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,297 @@
+    # This is the real magic soup of the class, we catch anything that isn't strictly defined for
+    # ourselves and pass it on to whatever client we've been given.
+    def __getattr__(self, name, *args, **kwargs):
+        # For no retry:
+        # return self.client.__getattribute__(name, *args, **kwargs)
```
I would just leave this comment out, since it doesn't add anything to the understanding.

cgordon commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,297 @@
+                    # If we're at the limit, just raise the exception to caller.
+                    # Exception type doesn't matter.
+                    raise exc
+                elif isinstance(exc, self.retry_for) or not isinstance(exc, self.do_not_retry_for):
```
There are four cases here:

(1) exc in self.retry_for, exc NOT in self_do_not_retry_for: this case is correct, and will retry

(2) exc in self.retry_for, exc in self.do_not_retry_for: this will retry the exception, and this is probably a user error, maybe check that users don't do this as a precondition in __init__?

(3) exc NOT in self.retry_for, exc in self.do_not_retry_for: this case is correct, and will not retry

(4) exc NOT in self.retry_for, exc NOT in self.do_not_retry_for: this case is sort of questionable. It will retry the exception, even though the user didn't explicitly ask for it to be retried. This is where we need a more clear contract for what users can expect from this logic.

martinnj replied:
This one will be clarified with documentation. Case 4 is a little dicey yes.
My intuition was that if you haven't  asked specified any go/no-go exceptions, revert to simpler behavior: Everything is just getting retried.

cgordon commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,297 @@
+                return self.client.__getattribute__(name, *args, **kwargs)
+            except Exception as exc:
+
+                if attempt == self.attempts:
```
I think you can simplify this conditional block as:

```
if attempt == self.attempts or not isinstance(exc, self._retry_for) or isinstance(exc, self.do_not_retry_for):
    raise exc

sleep(self.retry_delay)
```

martinnj replied:
Nice catch. Thanks. 👍

cgordon commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,297 @@
+        return self.delete(key, noreply=True)
+
+
+    # "static" implementations.
```
You can just delete the comments below this line and stick with the dynamic implementation.

cgordon reviewed (changes requested):
There are a couple existing "retry" libraries on PyPI, and it is probably worth either implementing their full set of functionality, or just using them directly. Here are the two I found:

https://pypi.org/project/retry/
https://pypi.org/project/retrying/

martinnj commented:
> 
> 
> There are a couple existing "retry" libraries on PyPI, and it is probably worth either implementing their full set of functionality, or just using them directly. Here are the two I found:
> 
> https://pypi.org/project/retry/
> https://pypi.org/project/retrying/

Was there any specific features from those libraries you where wanting?
Currently the timeout, retry count and exception based actions is implemented.

I didn't look for existing libraries because adding dependencies for something like this would be a shame in my eyes.

4383 commented:
> There are a couple existing "retry" libraries on PyPI, and it is probably worth either implementing their full set of functionality, or just using them directly. Here are the two I found:
> 
> https://pypi.org/project/retry/
> https://pypi.org/project/retrying/

Tenacity is an interesting implementation and already well used on openstack.

However, I agree with @martinnj , if possible, it would be more interesting to keep pymemcache as a pure python library (I voluntarly ignore `six`). Maybe by only providing specific retry strategies without tons of configurations. Less is more.

4383 commented:
Also notice that Wednesday I also started to implement similar features, however, I didn't yet proposed them before today  through a pull request.  #326

By introspection the passed clients, I dynamically decorate their methods to make them "retryable".

It's not a problem for me if you decide to continue with this version (that of @martinnj). Maybe we could converge our implementations to design and co-author something, it's up to you to decide. I propose you dispose.

martinnj commented:
> It's not a problem for me if you decide to continue with this version (that of @martinnj). Maybe we could converge our implementations to design and co-author something, it's up to you to decide. I propose you dispose.

Wild timing, I'll just wait with spending more time until a collaborator decides either way. :)

martinnj commented:
I had some time so I thought I'd give it a shot anyway.

Changes:
- I had to redo the retry/passthrough logic since I had grossly misunderstood `__getattr__`.
- Fixed style issues.
- Fleshed out docstrings.
- Added unit tests.

I haven't added a unit-test for the "delay" function. If forced I would time how long the call took and check that it was strictly longer than the delay argument. But I also feel like inserting second long delays in the unit-tests is a bit weird.

cgordon commented:
@martinnj I'm out of town right now and mostly away from my computer. I will have time to look through this next week, apologies for the delay.

martinnj commented:
> 
> 
> @martinnj I'm out of town right now and mostly away from my computer. I will have time to look through this next week, apologies for the delay.

No stress, it's all good.

4383 commented:
Hello

@martinnj: Please can you also add related doc with some examples of usage? 
maybe there https://pymemcache.readthedocs.io/en/latest/getting_started.html

4383 commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,183 @@
+
+    # Ensure the argument is a tuple, or raise an error if we can't convert it.
+    if argument is None:
+        return tuple()
```
This could be the default value of `argument_tuple` and the condition block could be reduced to if/elif/else:

Example:
```
if isinstance(argument, tuple):
        argument_tuple = argument
elif isinstance(argument, Iterable) and \
            not isinstance(argument, (str, bytes, bytearray)):
        argument_tuple = tuple(argument)
else:
        raise ValueError("%s was not a tuple or Iterable." % argument_name)
```

Anyway, I don't want to nitpick our current implementation, this code LGTM in the current state.

4383 replied:
This check can remain. It will be complementary with testing list and tuple. See my next comments.

martinnj replied:
Went ahead and simplified the check as discussed below.

4383 reviewed (changes requested):
Hello see my inline comment about abc

4383 commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,183 @@
+          *args: args, array arguments to pass to the function.
+          **kwargs: kwargs, keyword arguments to pass to the function.
+        """
+        for attempt in range(1, self._attempts + 1):
```
Can't we start our range from `0` to `self._attempts` to avoid to bump `self._attempts`?

By example if user pass `attempts=1` that will avoid a mental gymnastic to get our starting point during code reading. You protected the execution from values inferior to `1`, so I don't think that we need to tweak it during runtime.

martinnj replied:
Sure. I'm am not opinionated on it.
https://github.com/martinnj/pymemcache/commit/b0782c595dafa92c1608abd8f749c255ebf71edd

4383 commented on pymemcache/test/test_client_retry.py:
```diff
@@ -0,0 +1,315 @@
+
+import functools
```
From a pep8 point of view I think that you could group the imports by lib types (stdlib first, then third libraries, and then the current lib)

martinnj replied:
You're right. Didn't clean up properly after myself when I was writing tests.
https://github.com/martinnj/pymemcache/commit/03d7c48fba36934659fc37f1a4844da46d00f681

martinnj commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,183 @@
+""" Module containing the RetryingClient wrapper class. """
+
+from collections import Iterable
```
The library currently supports Python 2.7 which doesn't have `collections.abc`, but I don't see any reason why it can't be done with a `try .. except`. 👍

4383 replied:
Oh yes good catch. My bad.
I was thinking that the py2 support was already dropped, but indeed that's not the case yet

https://github.com/pinterest/pymemcache/pull/321

martinnj replied:
Import became unnecessary after the check simplification below, so just removed it. Nicely spotted.

martinnj commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,183 @@
+        return tuple()
+    elif isinstance(argument, tuple):
+        argument_tuple = argument
+    elif isinstance(argument, Iterable) and \
```
Yeah, simplified the check. 
https://github.com/martinnj/pymemcache/commit/5da40908d0c9628e6185ca119bcd98afa65e279c

4383 commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,178 @@
+    if argument is None:
+        return tuple()
+    elif not isinstance(argument, (tuple, set, list)):
+        raise ValueError("%s was not a tuple or Iterable." % argument_name)
```
Sorry I didn't noticed this exception message during my previous review.

What do you think about a message like this:

> An error occur with your passed argument. This method accept three possible kind of sequence type (tuple, set, list) for the argument "argument". If you see this message that mean that your argument isn't valide. Your argument type is {type(argument)}. If you continue to see this message, then, please check your passed arguments.

Maybe we could rename "argument" as "exceptions" and rename "argument_name" as "exception_name".
In this context I think that we could speak about exception rather than seeking to be too generalist. That will make the code more human readable, explicit and semantic.

martinnj replied:
I don't know that I agree with having an exception message that long. Makes it seem like the information should be in the client docstring (both the check method and the constructor) instead.

I agree that the wording could perhaps be better, and I also see that the exception message doesn't actually match the types we check for.

My reasoning for the variable naming is like so:
- `argument_name` is the name of the argument we're currently checking. In itself it doesn't have anything to do with an exception, it's just for giving sensible `ValueError` messages, it will tell whoever catches it exactly which argument (`retry_for` or `do_not_retry_for`) contained an invalid value.
- `argument` is the value we're checking. I'm renaming it to `argument_value` since this is more explicit.

Done in https://github.com/martinnj/pymemcache/commit/6bd21d0261f57d54c8c3d1211088bc27f07f5ee3

4383 replied:
No problem, the current version LGTM, that was more a suggestion. Leave that as it is if you think that it's ok.

4383 commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,178 @@
+    # (Catchable)
+    if not all([issubclass(arg, Exception) for arg in argument_tuple]):
+        raise ValueError(
+            "%s contained elements other than Exceptions." % argument_name
```
Some kind of remark here.

The error message could be more detailed and it could bring more details about why the users face it and what are the solutions for them. We should be proactive and lead users on the right path to avoid mistakes and so reduce the human support in a second time (github issues, etc...).

martinnj replied:
I'm still not sold on super wordy exception messages, so I'll let a collaborator weigh in on that one.
I've reworded the exception to be more specific about what is required though.
https://github.com/martinnj/pymemcache/commit/6bd21d0261f57d54c8c3d1211088bc27f07f5ee3

4383 commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,178 @@
+        for exc_class in self._retry_for:
+            if exc_class in self._do_not_retry_for:
+                raise ValueError(
+                    "%s was present in both `retry_for` and "
```
Same think here. We should define in the message why the users face it, what we except and what users should do in this case.

martinnj replied:
Somewhat done with https://github.com/martinnj/pymemcache/commit/d01efbb09cdb5bc4d3d62eb59df714256833c2bd

4383 commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,178 @@
+        """
+
+        if attempts < 1:
+            raise ValueError("`attempts` argument must be at least 1.")
```
here too

martinnj replied:
Somewhat done with https://github.com/martinnj/pymemcache/commit/d01efbb09cdb5bc4d3d62eb59df714256833c2bd

martinnj replied:
Also Somewhat done with https://github.com/martinnj/pymemcache/commit/d01efbb09cdb5bc4d3d62eb59df714256833c2bd

4383 reviewed (commented):
Nice changes, thank you for applying my previous comments.

Here is a couple of new suggestions, they aren't mandatory comments. I think that they can help to improve this feature.

4383 commented:
Also, I think that it could be worth to reword and to squash your commits into a all in one commit to avoid to spread the history of these changes.

Example:
```
$ git rebase -i HEAD~12
reword ee1aa4c Add first draft for RetryingClient.
squash 9325a3c Corrections, docstrings and fixes for RetryingClient.
squash 0700983 Stop retring AttributeError and style fixes.
squash eddbe50 Stylefixes for flake8.
squash 6664775 Minor fixes and docstrings for RetryingClient.
squash cf2b44d Add unit-tests for RetryingClient.
squash f08d348 Add RetryingClient example to the getting started docuementation.
squash 71f81bb More robust import of Iterable.
squash b0782c5 0 index retry range.
squash 03d7c48 Reorder imports of tests for pop8ness.
squash eee78cc Sharpen test-case.
squash 5da4090 Simplify argument parsing.
```

All these changes are related to a single one feature.

martinnj commented on docs/getting_started.rst:
```diff
@@ -77,6 +77,34 @@ on if a server goes down.
 
+Using a built-in retrying
+-------------------------
+The library comes with a retrying client that allows you to retry calls that
```
I'm not strongly opinionated on documentation. In my experience if someone who didn't write the code tells you that a documentation part is easier to understand in a different way, they're usually right. :)
Done in: https://github.com/martinnj/pymemcache/commit/8c6013e04c7410430d06186cc7b0e92b4486c14d

martinnj commented:
> 
> 
> Also, I think that it could be worth to reword and to squash your commits into a all in one commit to avoid to spread the history of these changes.
> 
> Example:
> 
> ```
> $ git rebase -i HEAD~12
> reword ee1aa4c Add first draft for RetryingClient.
> squash 9325a3c Corrections, docstrings and fixes for RetryingClient.
> squash 0700983 Stop retring AttributeError and style fixes.
> squash eddbe50 Stylefixes for flake8.
> squash 6664775 Minor fixes and docstrings for RetryingClient.
> squash cf2b44d Add unit-tests for RetryingClient.
> squash f08d348 Add RetryingClient example to the getting started docuementation.
> squash 71f81bb More robust import of Iterable.
> squash b0782c5 0 index retry range.
> squash 03d7c48 Reorder imports of tests for pop8ness.
> squash eee78cc Sharpen test-case.
> squash 5da4090 Simplify argument parsing.
> ```
> 
> All these changes are related to a single one feature.

Plan was to squash just before merging, since the commits give context to some of the comments in the PR.

4383 commented:
> > Also, I think that it could be worth to reword and to squash your commits into a all in one commit to avoid to spread the history of these changes.
> > Example:
> > ```
> > $ git rebase -i HEAD~12
> > reword ee1aa4c Add first draft for RetryingClient.
> > squash 9325a3c Corrections, docstrings and fixes for RetryingClient.
> > squash 0700983 Stop retring AttributeError and style fixes.
> > squash eddbe50 Stylefixes for flake8.
> > squash 6664775 Minor fixes and docstrings for RetryingClient.
> > squash cf2b44d Add unit-tests for RetryingClient.
> > squash f08d348 Add RetryingClient example to the getting started docuementation.
> > squash 71f81bb More robust import of Iterable.
> > squash b0782c5 0 index retry range.
> > squash 03d7c48 Reorder imports of tests for pop8ness.
> > squash eee78cc Sharpen test-case.
> > squash 5da4090 Simplify argument parsing.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > All these changes are related to a single one feature.
> 
> Plan was to squash just before merging, since the commits give context to some of the comments in the PR.
> 
> Edit:
> Thanks for all the feedback btw. :)

Thanks to you too :)

No problem it make sense to squash it just before.

martinnj commented:
@cgordon Hate to be pushy, but any news on this one? :)

cgordon commented on pymemcache/client/retrying.py:
```diff
@@ -0,0 +1,185 @@
+            each attempt.
+            Defaults to 0.
+
+          retry_for: optional None|tuple|set|list, what exceptions to
+            allow retries for. Will allow retries for all exceptions if None.
+            Example:
+                `(MemcacheClientError, MemcacheUnexpectedCloseError)`
+            Accepts any class that is a subclass of Exception.
+            Defaults to None.
+
+          do_not_retry_for: optional None|tuple|set|list, what
+            exceptions should be retried. Will not block retries for any
+            Exception if None.
+            Example:
+                `(IOError, MemcacheIllegalInputError)`
+            Accepts any class that is a subclass of Exception.
+            Defaults to None.
```
I think it would be nicer to take a function that takes an `Exception` as an argument and returns `True` to retry or `False` to stop retries. That will give clients a lot more flexibility, and make the logic a little more straightforward. This is a subjective thing, so I won't hold up the commit on it.

martinnj replied:
I agree that passing something callable in would make it more flexible. But ff you are not strongly opinionated on this I would like to keep it like this, since it makes invocation simpler.

It might be an feature to add later though. I don't see them having to exclude each other, as long as any priorities in the checks is in the documentation.

cgordon reviewed (commented):
LGTM! I left one comment that you can do, or not, at your discretion.

cgordon commented:
Hmm, I hit "Comment" instead of "Approve" by accident, anyone know how to navigate this UI to approve this PR?

martinnj commented:
> 
> 
> Hmm, I hit "Comment" instead of "Approve" by accident, anyone know how to navigate this UI to approve this PR?

Good question. No clue. I've hit "Re-request review" for the blocking changes, might give you a prompt to clear it. :)
